### PR TITLE
En 9304/sender in out transfer

### PR DESCRIPTION
**Pull request by sasurobert**
Added and treated sender in output transfer

### CODE REVIEW

File: process/smartContract/process.go [hunk 1]
```diff
@@ -1785,9 +1789,15 @@ func (sc *scProcessor) createSmartContractResults(
 			result.GasLimit, _ = core.SafeAddUint64(result.GasLimit, vmOutput.GasRemaining)
 		}
 
+		if sc.flagSenderInOutTransfer.IsSet() && len(outputTransfer.SenderAddress) == len(tx.GetSndAddr()) &&
```
Comment (iulianpascalau) [inline process/smartContract/process.go:1792]: maybe extract this in a bool var?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: process/smartContract/builtInFunctions/saveUserName.go [hunk 2]
```diff
@@ -85,11 +85,12 @@ func (s *saveUserName) ProcessBuiltinFunction(
...
-			GasLimit:  vmInput.GasProvided,
-			GasLocked: vmInput.GasLocked,
-			Data:      []byte(setUserNameTxData),
-			CallType:  vmcommon.AsynchronousCall,
+			Value:         big.NewInt(0),
+			GasLimit:      vmInput.GasProvided,
+			GasLocked:     vmInput.GasLocked,
+			Data:          []byte(setUserNameTxData),
+			CallType:      vmcommon.AsynchronousCall,
+			SenderAddress: vmInput.CallerAddr,
```
Comment (raduchis) [inline process/smartContract/builtInFunctions/saveUserName.go:93]: no need for backwardsCompatibility flag here?

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go [hunk 3]
```diff
@@ -69,10 +69,11 @@ func (c *claimDeveloperRewards) ProcessBuiltinFunction(
...
 	outTransfer := vmcommon.OutputTransfer{
-		Value:    big.NewInt(0).Set(value),
-		GasLimit: 0,
-		Data:     nil,
-		CallType: vmcommon.DirectCall,
+		Value:         big.NewInt(0).Set(value),
+		GasLimit:      0,
+		Data:          nil,
+		CallType:      vmcommon.DirectCall,
+		SenderAddress: vmInput.CallerAddr,
```
Comment (raduchis) [inline process/smartContract/builtInFunctions/claimDeveloperRewards.go:76]: no need for backwardsCompatibility flag here?

Comment (sasurobert) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:76]: it is already put inside smart contract processor.

File: process/smartContract/builtInFunctions/saveUserName.go (hunk 2 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/saveUserName.go:93]: it is already put inside smart contract processor.

File: cmd/node/config/config.toml [hunk 4]
```diff
@@ -58,6 +58,9 @@
    # ArwenESDTFunctionsEnableEpoch represents the epoch when the new ESDT functions are enabled in arwen
    ArwenESDTFunctionsEnableEpoch = 5
 
+   # SenderInOutTransferEnableEpoch represents the epoch when we enabled different senders in output transfer
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/config/config.toml:61]: done

File: data/vm/vmOutputApi.go [hunk 5]
```diff
@@ -42,10 +42,11 @@ type OutputAccountApi struct {
...
 type OutputTransferApi struct {
-	Value    *big.Int          `json:"value"`
-	GasLimit uint64            `json:"gasLimit"`
-	Data     []byte            `json:"data"`
-	CallType vmcommon.CallType `json:"callType"`
+	Value         *big.Int          `json:"value"`
+	GasLimit      uint64            `json:"gasLimit"`
+	Data          []byte            `json:"data"`
+	CallType      vmcommon.CallType `json:"callType"`
+	SenderAddress []byte            `json:"senderAddress"`
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/vm/vmOutputApi.go:49]: done

File: process/smartContract/process.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/process.go:1792]: done

**Review by bogdan-rosianu — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**